DOCSP-12448: Backend Phase 2 Tutorial (#499)
mongodb/docs-realm: source/tutorial/realm-app.txt

.. default-domain:: mongodb

Overview
--------

Before we can implement any client SDK functionality for the Task
Tracker, we need to create a backend {+app+} to serve SDK requests
like authentication and {+sync+}. To do this, you'll have to:

1. Create an Atlas account.
#. Create a free Atlas cluster.
#. Install the Realm CLI.
#. Add a programmatic API key to your Atlas account.
#. Log into the Realm CLI with your API key.
#. Download the Task Tracker backend configuration.
#. Use the Realm CLI to create a new Task Tracker backend Realm App.

Once we've finished these steps, we can start writing code to implement

> a frontend using iOS, Android, or JavaScript.

Reviewer: I understand the motivation for writing the section like this, but I'm finding this list and page more daunting than it needs to be. Let's try rewriting this so that the emphasis is on the Realm parts with the key points that aren't covered by the Atlas sign-up flow called out. We could even move the detailed Atlas stuff (excellent quality) to another page so as not to frighten people away.

The key points are:
- We have prepared the backend Realm app for you, complete with the functions, triggers, schemas, and sync configuration you'll need to connect with one of our front-end apps.
- You can import the backend app using the Realm CLI with an Atlas API key.
- You need an Atlas cluster running MongoDB 4.4 to store the data for your task tracker app. You can get started with a free Atlas cluster.

Then the important steps of Atlas:
- Make sure you build your cluster with MongoDB 4.4
- Make sure your cluster is in the right region

Then link out to this text in another document if needed. ("Follow our step-by-step guide to creating a free Atlas account, 4.4 cluster, etc. etc...")

Author: Reworked with some of this, but leaving more explicit steps for now -- we'll see if drew/ian/katherine have similar feedback, or if they like the fact that this tutorial doesn't expect any prior knowledge of any of our products.

an existing MongoDB Atlas Account, you can
:ref:`proceed to the next step
<tutorial-task-tracker-create-a-free-atlas-cluster>`. If you don't have
an Atlas account, follow the steps below to create one:

1. Navigate to the `MongoDB Atlas login page
<https://account.mongodb.com/account/register>`__.

#. Click :guilabel:`Login`.

#. Either enter a new set of user credentials or click the
   :guilabel:`Sign Up with Google` button.

#. Click the checkbox to accept the Privacy Policy and Atlas Terms &
   Conditions.

#. Click :guilabel:`Sign Up` to create your account.

#. Follow the prompts to create an organization and project in your
   Atlas account. You can use the default suggested names or enter
   your own.

   .. image:: /images/atlas-create-project-and-organization.png
      :alt: The Atlas UI after creating an organization and project.

When you finish creating your organization and project, you should end
up on a screen that prompts you to create an Atlas cluster:

.. image:: /images/atlas-create-cluster.png
   :alt: The Atlas UI after creating an organization and project.

.. see::

   For additional details on how to create an account,

>    see :atlas:`Create an Atlas Account </tutorial/create-atlas-account/`.

Reviewer: Broken link:
```suggestion
   see :atlas:`Create an Atlas Account </tutorial/create-atlas-account/>`.
```

your Atlas account. In this view, you'll see Atlas's progress as it
initializes your new cluster:

.. image:: /images/atlas-creating-cluster.png
   :alt: The Atlas UI after creating an organization and project.

.. see::

   For additional details on how to create a MongoDB Atlas cluster,
   see :atlas:`Deploy a Free Tier Cluster
   </tutorial/deploy-free-tier-cluster/>`.

.. _tutorial-task-tracker-install-realm-cli:

C. Install the Realm CLI
------------------------

Now that you've created a cluster to use as the data source for your
{+app+}, we need some way to create the app itself. In most cases you'd
use the Realm UI, which you can access through the Atlas UI. However,
for the purposes of this tutorial, we're going to use the :ref:`Realm
Command Line Interface <realm-cli>`, also known as ``realm-cli``.

We're using the Realm CLI because ``realm-cli`` allows you to manage
your {+app+}s programmatically using JSON configuration files instead of
the {+service-short+} UI. While you could manually set up all of the
authentication, sync, function, and trigger settings for the Task
Tracker tutorial, it's much easier to just import everything from our

> provided configuration with a single command using ``realm-cli``. Follow

Reviewer: This paragraph (after the previous) is quite wordy. Can we cut this down? The point is realm-cli allows you to import the app that we have prepared for you.

the instructions below to install the Realm CLI in your development
environment using either a package manager or the ``realm-cli`` binary:

.. include:: /includes/install-realm-cli.rst

> When you've finished installing the ``realm-cli``, you can run the

Reviewer: Nitpick: wordiness (often indicated by "you've")
```suggestion
After installing the ``realm-cli``, you can run the
```

#. Click the :guilabel:`Create API Key` button.

#. In the :guilabel:`Description` text box, enter
   "API Key for the MongoDB Realm CLI".

   .. image:: /images/atlas-create-api-key.png
      :alt: Creating an API Key in the Atlas UI.

#. In the :guilabel:`Project Permissions` dropdown, select
   "Project Owner" and unselect "Project Read Only".

   .. image:: /images/atlas-api-key-permissions.png
      :alt: Grant your API Key "Project Owner" permissions.

#. Copy your Public API Key and save it somewhere.

#. Click :guilabel:`Next`.

#. Copy your Private API Key and save it somewhere; after leaving this
   page, you will no longer be able to view it via the Realm UI.

#. Click the :guilabel:`Add Whitelist Entry` button.

#. Click :guilabel:`Use Current IP Address`.

#. Click :guilabel:`Save`.

#. When you have safely recorded your private API key, click
   :guilabel:`Done` to navigate back to the Project Access Manager page.

.. see::

   For additional details about how to create a programmatic API key
   in {+atlas+}, see the
   :atlas:`documentation for configuring API access
   </configure-api-access/#programmatic-api-keys>`.

.. _tutorial-task-tracker-login-to-realm-cli:

> E. Log into the Realm CLI with your API Key

Reviewer: Can we get away with combining this step with the previous to reduce the overall number of steps? These two steps are often linked together anyway (I always lose the old API key when I want to log in with the CLI again, anyway)

-------------------------------------------

Now that you've added a programmatic API key to your Atlas account,
you can use that programmatic API key to log into your account via the
Realm CLI. To log into your account using the programmatic API key,

> follow the instructions below:

Reviewer: If you take my previous suggestion to combine this step with the previous step, you can probably delete this whole paragraph.

F. Download the Task Tracker Backend Configuration
--------------------------------------------------

Now that you've got an Atlas account, an Atlas cluster running MongoDB
4.4, and an authenticated ``realm-cli`` session, you're ready to fetch
the Task Tracker backend configuration. Begin by downloading the
``realm-tutorial-backend`` github repository using the ``git`` command
line tool:

.. code-block:: console

   git clone git@github.com:mongodb-university/realm-tutorial-backend.git

You should see output like the following:

.. code-block:: console

   Cloning into 'realm-tutorial-backend'...
   remote: Enumerating objects: 39, done.
   remote: Counting objects: 100% (39/39), done.
   remote: Compressing objects: 100% (31/31), done.
   remote: Total 39 (delta 3), reused 39 (delta 3), pack-reused 0
   Receiving objects: 100% (39/39), 7.25 KiB | 1.81 MiB/s, done.
   Resolving deltas: 100% (3/3), done.

The directory where you ran the above ``git clone`` command should now
contain another directory called ``realm-tutorial-backend``. We'll use
the contents of that directory to create your very own Task Tracker
{+app+} backend.

.. _tutorial-task-tracker-create-backend:

> G. Use the Realm CLI to Create a New Task Tracker Backend Realm App

Reviewer: Again, can we combine these two steps? Clone & Import

-------------------------------------------------------------------

Finally, we'll use the Realm CLI to create a new {+app+} based on the
configuration you just downloaded. Follow the instructions below to
create your app:

1. Navigate to the ``tasktracker`` directory of the
   ``realm-tutorial-backend`` project:

   .. code-block:: console

      cd realm-tutorial-backend/tasktracker/

#. Run the ``import`` command of the ``realm-cli`` to create your
   app.

   .. code-block:: console

      realm-cli import

#. ``realm-cli`` may take a few seconds to query your Atlas project, but
   you should soon see the following output:

   .. code-block:: console

      this app does not exist yet: would you like to create a new app? [y/n]:

   Press "y", then press ENTER to confirm your intention to create a new app.

#. ``realm-cli`` will prompt you for the following details:

   .. code-block:: console

      App name [tasktracker]:
      Available Projects:
      Project 0 - <project ID>
      Atlas Project Name or ID [Project 0]:
      Location [US-VA]:
      Deployment Model [GLOBAL]:

>     Press ENTER at each prompt to use the default value for your app

Reviewer: The indentation here makes this sentence a part of the code-block.

    configuration.

If you see output like the following, your app has been successfully
created:

.. code-block:: console

>    Successfully imported 'tasktracker-mrksq'

Reviewer: Perhaps worth mentioning that this is the elusive "Realm app ID" you'll be pasting around.

You can also confirm that your app was created successfully by
navigating to the :guilabel:`Realm` tab in the Atlas UI. You should
see a {+app+} named ``tasktracker``:

.. image:: /images/realm-app-tasktracker.png
   :alt: A newly created TaskTracker backend in the Atlas UI.

.. _tutorial-task-tracker-verify-configuration:

H. Verify that the Task Tracker Backend is Properly Configured
--------------------------------------------------------------

> Now that you've successfully created your application, it's time to

Reviewer: Maybe at this point we can allow eager people to go straight to the client steps?

- remove a team member from a user's project

- reset a user's password (unimplemented)

Click on any of the functions to view the JavaScript function
definition. You can even run the function with test input.

.. see::

   To learn more about functions in {+service+}, see the
   :ref:`documentation on functions <functions>`.

Triggers
~~~~~~~~

In the :guilabel:`Triggers` section, you'll find information about
**triggers**, certain criteria that, when met, execute logic in the
Task Tracker backend via a function. In Task Tracker, you'll find one
trigger: ``onNewUser``. Click on the trigger to see details about the
type of trigger and the function that that trigger activates.
``onNewUser`` has just one purpose: when a new user creates an account
(hence the "Create" action type and the "Authentication" trigger type),
``onNewUser`` calls the ``createNewUserDocument`` function which
initializes the custom user data for that user. This provides the first
data needed to create that user's personal project, and allows them to
start adding to-do tasks or even other users as project members.

.. see::

   To learn more about triggers in {+service+}, see the

>    :ref:`documentation on triggers <triggers>`.

Reviewer: Under "What's Next" be sure to add the Node tutorial!
